### PR TITLE
WIP: Update url shortener sample

### PR DESCRIPTION
**Pull request by michaelbausor**
cc @sirdarckcat

Addressing issues raised in https://github.com/google/google-api-php-client/pull/1197 (b/37360079)

### CODE REVIEW

File: examples/url-shortener.php [hunk 1]
```diff
@@ -101,12 +110,30 @@
...
+    $_SESSION['access_token'] = $client->getAccessToken();
+  } else {
+    $invalidCsrf = true;
+    generateCsrfToken();
+  }
+}
+
+function generateCsrfToken()
+{
+  $_SESSION['csrf_token'] = bin2hex(openssl_random_pseudo_bytes(32));
```
Comment (sirdarckcat) [inline examples/url-shortener.php]: you need to check the return value of the second parameter
http://php.net/manual/en/function.openssl-random-pseudo-bytes.php

Comment (michaelbausor) [inline examples/url-shortener.php]: I am not sure what behaviour we want here in the case that the $crypto_strong parameter comes back false. Should we emit a warning? It seems wrong to fail completely, as that is not a great experience for users trying to test out the example code on their local machine.

File: examples/url-shortener.php [hunk 2]
```diff
@@ -101,12 +110,30 @@
...
+}
+
+function validateCsrfToken()
```
Comment (bshaffer) [inline examples/url-shortener.php]: This is great! However, we have a place for helper functions like this in [`examples/templates/base.php`](https://github.com/google/google-api-php-client/blob/master/examples/templates/base.php). Please move this function there.

File: examples/url-shortener.php [hunk 3]
```diff
@@ -101,12 +110,30 @@
...
-  $_SESSION['access_token'] = $client->getAccessToken();
+if ($client->getAccessToken() && isset($_REQUEST['url'])) {
+  if (validateCsrfToken()) {
+    $url = new Google_Service_Urlshortener_Url();
+    $url->longUrl = $_REQUEST['url'];
+    $short = $service->url->insert($url);
+    $_SESSION['access_token'] = $client->getAccessToken();
+  } else {
+    $invalidCsrf = true;
+    generateCsrfToken();
```
Comment (bshaffer) [inline examples/url-shortener.php]: We can just follow the model of the [other validation functions](https://github.com/michaelbausor/google-api-php-client/blob/5e085f1041467a5ffacf17663911bddceca51f22/examples/url-shortener.php#L26) if this fails:

```php
if (!validateCsrfToken()) {
  echo invalidCsrfTokenWarning();
  return;
}

// and in base.php
function invalidCsrfTokenWarning()
{
    return 'The CSRF token is invalid, your session probably expired. Please refresh the page ';
}
```

File: examples/url-shortener.php [hunk 4]
```diff
@@ -115,18 +142,28 @@
...
 <?php else: ?>
   You created a short link! <br />
   <a href="<?= $short['id'] ?>"><?= $short['id'] ?></a>
   <div class="shortened">
     <pre><?php var_export($short) ?></pre>
   </div>
-  <a href="<?= $_SERVER['PHP_SELF'] ?>">Create another</a>
+  <a href="<?= htmlspecialchars($_SERVER['PHP_SELF']); ?>">Create another</a>
+<?php endif ?>
+<?php if (isset($invalidCsrf) && $invalidCsrf): ?>
```
Comment (bshaffer) [inline examples/url-shortener.php]: See above comment and we can get rid of this block.

File: examples/url-shortener.php [hunk 5]
```diff
@@ -115,18 +142,28 @@
     <a class='login' href='<?= $authUrl ?>'>Connect Me!</a>
   </div>
 <?php elseif (empty($short)): ?>
-  <form id="url" method="GET" action="<?= $_SERVER['PHP_SELF'] ?>">
+  <form id="url" method="POST" action="<?= htmlspecialchars($_SERVER['PHP_SELF']); ?>">
     <input name="url" class="url" type="text">
+    <input type="hidden" name="csrf_token" value="<?php echo $_SESSION['csrf_token']; ?>" />
```
Comment (bshaffer) [inline examples/url-shortener.php]: Let's make a new function `getCsrfToken` that generates this if it doesn't exist.

File: examples/url-shortener.php [hunk 6]
```diff
@@ -60,7 +60,13 @@
  * local access token in this case
  ************************************************/
 if (isset($_REQUEST['logout'])) {
-  unset($_SESSION['access_token']);
+  if (validateCsrfToken()) {
```
Comment (bshaffer) [inline examples/url-shortener.php]: is any of this necessary? I don't understand why we need to remove the csrf token on logout, or why the validate function needs to be called. At the very least we can just unset the token.

**Comment by bshaffer** [comment]
closing in favor of #1240

**Pull request closed without merging**